bazaarvoice/emodb
# EMO-6249: Introduced basic event tracing for replay and move databus operations

## Github Issue #

None

## What Are We Doing Here?

There have been reports that after a databus replay and move operation a small number of events which should have been posted to a client's databus subscription are not present.  The client has logged all databus events which they received and a comparative analysis shows that several records that were updated within the window of the replay request did not appear in the final subscription after the replay and move.  However, it is still unknown if the issue was on the client's side, with the client presumably discarding events prior to client logging call, or on Emo's side due to some squirrelly difficult-to-reproduce error.  Specifically, there is no way to definitively compare the client's log to what was actually replayed and moved by Emo, making it impossible to resolve the issue or even know which system requires further investigation.

This PR adds the ability to trace every event replayed or moved in one of those operations and dump the log to a CSV file in the blob store.  For example:

```csv
7041a0ef-13ef-11e7-a105-2eb4b6c28252,3/28/17 2:47 PM,sample_table,record-995,DedupQueue#moveToRawChannel
7041a0f1-13ef-11e7-a105-2eb4b6c28252,3/28/17 2:47 PM,sample_table,record-996,DedupQueue#moveToRawChannel
7041a0f3-13ef-11e7-a105-2eb4b6c28252,3/28/17 2:47 PM,sample_table,record-997,DedupQueue#moveToRawChannel
7041a0f5-13ef-11e7-a105-2eb4b6c28252,3/28/17 2:47 PM,sample_table,record-998,DedupQueue#moveToRawChannel
7041a0f7-13ef-11e7-a105-2eb4b6c28252,3/28/17 2:47 PM,sample_table,record-999,DedupQueue#moveToRawChannel
```

It's possible to extend this to support other logging options, such as writing a file to S3 or just plain log4j, but for now a blob document seemed like a good compromise for providing a publicly-accessible trace dump without adding any external dependencies or relying on an Emo admin's intervention, such as would be needed if the log were written to a local file on the Emo instance.
## How to Test and Verify

1. Check out this PR
2. Create a blob table called "movetrace"
3. Create a subscription "sourceSub" with condition `alwaysTrue()`
4. Create at least one SoR table and add multiple documents to it.
5. Create another subscription called "destSub", also with condition `alwaysTrue()`
6. Perform a move operation with a trace parameter, such as:

   `POST /bus/1/_move?from=sourceSub&to=destSub&tracer=type:blob,table:movetrace,blobId:trace1,ttl:1209600000,gzipped:!t`

7. Get the CSV blob from "movetrace/trace1" and un-gzip it
8. Verify all documents which should have been in `sourceSub` were moved and a record of each is in the CSV.
9. Repeat with a replay operation

## Risk

Low.  I believe all of the locations where events are operated on in move and replay are included, but even if wrong there's no additional risk.  The move and replay operations themselves will continue to work as they did before this change.  If any errors of omission in the logs are found they can be resolved in future updates.  Additionally, all calls to the tracer are protected so even if an exception were thrown by the tracer it shouldn't affect the actual operation.

### Level 

`Medium`

### Required Testing

`Manual`

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
